// @arco-design/web-vue/lib/descriptions/descriptions-item.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var context = require("./context.js");
var useIndex = require("../_hooks/use-index.js");
var pluginVue_exportHelper = require("../_virtual/plugin-vue_export-helper.js");
const _sfc_main = vue.defineComponent({
  name: "DescriptionsItem",
  props: {
    span: {
      type: Number,
      default: 1
    },
    label: String
  },
  setup(props) {
    var _a;
    const { span } = vue.toRefs(props);
    const prefixCls = globalConfig.getPrefixCls("descriptions");
    const descCtx = vue.inject(context.descriptionsInjectionKey, {});
    const instance = vue.getCurrentInstance();
    const itemRef = vue.ref();
    const { computedIndex } = useIndex.useIndex({
      itemRef,
      selector: `.${prefixCls}-item-value`,
      parentClassName: `${prefixCls}-table`
    });
    const itemData = vue.reactive({
      index: computedIndex,
      span
    });
    if (instance == null ? void 0 : instance.uid) {
      (_a = descCtx.addItem) == null ? void 0 : _a.call(descCtx, instance.uid, itemData);
    }
    const getItemRef = () => {
      var _a2;
      const itemEle = ((_a2 = instance == null ? void 0 : instance.proxy) == null ? void 0 : _a2.$el).parentElement;
      if (itemEle && itemEle !== itemRef.value) {
        itemRef.value = itemEle;
      }
    };
    vue.onMounted(() => getItemRef());
    vue.onUpdated(() => getItemRef());
    vue.onBeforeUnmount(() => {
      var _a2;
      if (instance == null ? void 0 : instance.uid) {
        (_a2 = descCtx.removeItem) == null ? void 0 : _a2.call(descCtx, instance.uid);
      }
    });
    return {
      prefixCls
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return vue.renderSlot(_ctx.$slots, "default");
}
var DescriptionsItem = /* @__PURE__ */ pluginVue_exportHelper(_sfc_main, [["render", _sfc_render]]);
module.exports = DescriptionsItem;
